cli/lock.ml

```diff
@@ -189,25 +192,72 @@ let display_verbose_diagnostics = function
   | None -> false
   | Some l -> l >= Logs.Info
 
-let interpret_solver_error ~switch_state = function
+let interpret_solver_error ~repositories solver = function
   | `Msg _ as err -> err
   | `Diagnostics d ->
-      (match Opam_solve.not_buildable_with_dune d with
+      (match Opam_solve.not_buildable_with_dune solver d with
       | [] -> ()
       | offending_packages ->
-          check_dune_universe_repo ~switch_state offending_packages);
+          check_dune_universe_repo ~repositories offending_packages);
       let verbose = display_verbose_diagnostics (Logs.level ()) in
-      Opam_solve.diagnostics_message ~verbose d
+      Opam_solve.diagnostics_message ~verbose solver d
 
-let calculate_opam ~build_only ~allow_jbuilder ~local_opam_files ~ocaml_version
-    ~target_packages =
+let pull_repository url =
+  match (OpamUrl.local_dir url, url.backend) with
+  | Some path, `rsync ->
+      Ok (OpamFilename.Dir.to_string OpamFilename.Op.(path / "packages"))
+  | _ ->
+      (* TODO before release *)
+      Rresult.R.error_msg
+        "Only non git, local filesystem URLs (file://) are supported at the \
+         moment"
+
+let pull_repositories repositories =
+  Result.List.map ~f:pull_repository repositories
+
+let extract_opam_env ~source_config global_state =
+  match (source_config : Source_opam_file.config) with
+  | { global_vars = Some env; _ } -> env
+  | { global_vars = None; _ } ->
+      let vars = global_state.OpamStateTypes.global_variables in
+      OpamVariable.Map.fold
+        (fun var (lazy_content, _doc) acc ->
+          let name = OpamVariable.to_string var in
+          match Lazy.force lazy_content with
+          | None -> acc
+          | Some content -> String.Map.add ~key:name ~data:content acc)
+        vars String.Map.empty
+
+let calculate_opam ~source_config ~build_only ~allow_jbuilder ~local_opam_files
+    ~ocaml_version ~target_packages =
   let open Result.O in
   OpamGlobalState.with_ `Lock_none (fun global_state ->
-      OpamSwitchState.with_ `Lock_none global_state (fun switch_state ->
-          let* pin_depends = get_pin_depends ~global_state local_opam_files in
+      let* pin_depends = get_pin_depends ~global_state local_opam_files in
+      match (source_config : Source_opam_file.config) with
+      | { repositories = Some repositories; _ } ->
+          let repositories = OpamUrl.Set.elements repositories in
+          Logs.info (fun l ->
+              l "Solve using explicit repositories:\n%a"
+                Fmt.(list ~sep:(const char '\n') Opam.Pp.url)
```

Comment:
Honestly a list of URLs that are usually long will be hard to read on a single line, it will be split into several lines either way so I think here it makes sense to split it, to give you an idea, here's how it looks like:
```
opam-monorepo: [DEBUG] Pulling pin depends: opam-0install.dev
opam-monorepo: [INFO] Solve using explicit repositories:
file:///home/nathan/code/opam-overlays
file:///home/nathan/code/opam-repository
==> Found 51 opam dependencies for the target package.
```
I think it's okay but if you feel like we shouldn't do it I'm open to suggestions! Eventually we could use a more `Format` friendly line break indicator to preserve indentation, I went with the quick and dirty `\n` here out of laziness.

Reply:
Also: It might be equivalent to `any '\n'`, I find the `any` name highly confusing which made me unconsciously avoid it I think, happy to simplify this!

Reply:
Yeah, you might be right, having a newline separated list of URLs might be better for the readability overall.

I tried building some demo code to implement breaking with indentation with `Fmt` using `Fmt.(v)box` and `Fmt.cut` but for some reason it never wants to properly indent. I think I need to read on how the formatting boxes work (which would probably also help understand why in general we seem to break text way too early).